src/pykeen/models/base.py

```diff
@@ -32,27 +33,253 @@
+        :raises RuntimeError:
+            If the chosen loss function does not allow the calculation of margin ranking
+        :return: dtype: float, scalar
+            The margin ranking loss value.
+        """
+
+    @abstractmethod
+    def compute_label_loss(
+        self,
+        predictions: torch.FloatTensor,
+        labels: torch.FloatTensor,
+    ) -> torch.FloatTensor:
+        """Compute the classification loss.
+
+        :param predictions: shape: s
+            The tensor containing predictions.
+        :param labels: shape: s
+            The tensor containing labels.
+
+        :return: dtype: float, scalar
+            The label loss value.
+        """
+
+    @abstractmethod
+    def compute_self_adversarial_negative_sampling_loss(
+        self,
+        positive_scores: torch.FloatTensor,
+        negative_scores: torch.FloatTensor,
+    ) -> torch.FloatTensor:
+        """Compute self adversarial negative sampling loss.
+
+        :param positive_scores: shape: s
+            The tensor containing the positive scores.
+        :param negative_scores: shape: s
+            Tensor containing the negative scores.
+        :raises RuntimeError:
+            If the chosen loss does not allow the calculation of self adversarial negative sampling losses.
+        :return: dtype: float, scalar
+            The loss value.
+        """
+
+    """Concrete methods"""
+
+    def _set_device(self, device: DeviceHint = None) -> None:
+        """Set the Torch device to use."""
+        self.device = resolve_device(device=device)
+
+    def to_cpu_(self) -> Model:
+        """Transfer the entire model to CPU."""
+        self._set_device('cpu')
+        return self.to_device_()
+
+    def to_gpu_(self) -> Model:
+        """Transfer the entire model to GPU."""
+        self._set_device('cuda')
+        return self.to_device_()
+
+    def get_grad_params(self) -> Iterable[nn.Parameter]:
+        """Get the parameters that require gradients."""
+        # TODO: Why do we need that? The optimizer takes care of filtering the parameters.
+        return filter(lambda p: p.requires_grad, self.parameters())
+
+    @property
+    def num_parameter_bytes(self) -> int:
+        """Calculate the number of bytes used for all parameters of the model."""
+        return sum(
+            param.numel() * param.element_size()
+            for param in self.parameters(recurse=True)
+        )
+
+    def save_state(self, path: str) -> None:
+        """Save the state of the model.
+
+        :param path:
+            Path of the file where to store the state in.
+        """
+        torch.save(self.state_dict(), path)
+
+    def load_state(self, path: str) -> None:
+        """Load the state of the model.
+
+        :param path:
+            Path of the file where to load the state from.
+        """
+        self.load_state_dict(torch.load(path, map_location=self.device))
+
+
+class OModel(Model, ABC, autoreset=False):
```

Comment:
Is there a better name than `OModel`?

Reply:
🤷‍♀️ I'm open to suggestions

Reply:
Also this class name isn't really exposed to users ever since they can use `Model` for type annotations and the implementations for everything else

Reply:
`_AbstractModel` maybe?

Reply:
Done in f57d318e